# AMPnano, a command line tool for Automated Multi-file Processing of NGS data from Oxford nanopore thecnologies

## Description:
The Script performs an fully automated data processing of multiple fastq files generated by miniON (OXFORD NANOPORE TECKNOLOGIES). This tool uses a medaka network model that generates preliminary variants which is further refined down the pipeline. The tool was developed for the TELE-vir project and Instituto Nacional de Sa√∫de Doutor Ricardo Jorge (http://www.insa.min-saude.pt/). 

## The tool requires the following mandatory INPUTS:
	* -g  < the reference genome sequence path (must be in fasta format) >
	* -s  < the folder path where with the reads files are located (must be in fastq format) >
	* -i  < the path for the metadata file (*.tsv or *.csv ) containing samples ID and file names >

## As OUTPUTS, the script generates the following files organized in sample folders inside a results folder:
    *  Predicted consensus file with sample ID (consensus.fasta)
    *  Bam files
    *  Sample coverage file (depth.gz)
    *  Curated variants file (medaka_variants.vcf) with detail information on each detected variant
    *  Sample Coverage quality plot with variants location in the sequence locus
    *  Basic stats reports of the sample
    *  Overal analysis reports in csv files, one with the list of all mutations detected and another with main stats"
    *  Run report contaning pipeline parameters and other info

Optionally, you can get the paths using the choose GUI option by typing "choose" after each mandatory input command. Example:  -g choose -s choose -i choose.
You can also set most pipeline parameters for tuning the analysis by adding adicional arguments. If not defaut settings will be applied.


## Pre-instalation software requisites:
* anaconda python 3.7 distribution or later 
* mathplotlib
* biopython 
* anaconda python distribution 
* Medaka 1.2.1 (anaconda forge installation)
* Nanofilt 2.6.0  
* NanoStats 1.4.0 


## Running instructions:

### On the command line using anaconda environment, activate medaka environment by typing  

	Activate the medaka environment on anaconda by running the following commands.
 
    	(base) $ conda activate medaka
    

### On the command line, go to the path where the script is located and run the following command: 

	(medaka) $ python AMP_TELEVIR_CLI.py  -g <path reference> -s <path data> -i <path metadata>. 
      
 
 	or if you prefer to choose the location of files, just type choose instead of the path:
      
     
	(medaka) $ python SCRIPT_NAME.py -g choose -s choose -i choose. 
     
	In this case the file choose GUI will pop up and you can search you computer for the path to file. 


### If all works, you will see a folders for each sample and outputs files within will be generated systematically. 


